File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public void helper(int n, List<String> ans,String str , int open,int close){
        if(str.length()==n*2){
            ans.add(str);
            return;
        }
        if(open<n){
            helper(n,ans,str+'(',open+1,close);
        }
        if(close<open){
            helper(n,ans,str+')',open,close+1);
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> ans= new ArrayList<>();
        helper(n,ans,"",0,0);
        return ans;
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void helper(int n,vector<string>& ans,string curr,int o,int c){
        if(curr.size()==n*2){
            ans.push_back(curr);
            return;
        }
        if(o<n){
            helper(n,ans,curr+"(",o+1,c);
        }
        if(c<o){
            helper(n,ans,curr+")",o,c+1);
        }
    }
    vector<string> generateParenthesis(int n) {
        vector<string>ans;
        helper(n,ans,"(",1,0);
        return ans;
        
    }
};